generals/rewards/frequent_asset_reward_fn.py

```diff
@@ -0,0 +1,28 @@
+from generals.core.action import Action
+from generals.core.observation import Observation
+from generals.rewards.common import compute_num_cities_owned, compute_num_generals_owned, is_action_valid
+from generals.rewards.reward_fn import RewardFn
+
+
+class FrequentAssetRewardFn(RewardFn):
+    """This reward function is fairly frequent -- every action/turn should generate some kind of reward. And
+    it heavily takes into account the agent/players assets i.e. their land, army & cities.
+    """
+
+    def __call__(self, prior_obs: Observation, prior_action: Action, obs: Observation) -> float:
+        change_in_army_size = obs.owned_army_count - prior_obs.owned_army_count
+        change_in_land_owned = obs.owned_land_count - prior_obs.owned_land_count
+        change_in_num_cities_owned = compute_num_cities_owned(obs) - compute_num_cities_owned(prior_obs)
+        change_in_num_generals_owned = compute_num_generals_owned(obs) - compute_num_generals_owned(prior_obs)
+        # Moderately reward valid actions & penalize invalid actions.
+        valid_action_reward = 1 if is_action_valid(prior_action, prior_obs) else -5
+
+        reward = (
+            valid_action_reward
+            + 1 * change_in_army_size
+            + 5 * change_in_land_owned
+            + 10 * change_in_num_cities_owned
+            + 10_000 * change_in_num_generals_owned
```

Comment:
don't have an intuition on what `x10_000` does to gradients tho :D

Reply:
Hahaha yeah. For a vanilla gradient descent technique that may be problematic. But, for something like Adam which normalizes with an exponentially decaying running average of past updates I _think_ it might be ok? 

I wanted to choose a number high enough that even after a few hundred turns (i.e. $$\gamma^{300}$$) it'd dominate all other reward components. Right now my gamma/time-discount-factor is 0.995 so 10,000 may be a bit overkill!